### PR TITLE
#209 add OutOfSpecs Expression

### PR DESCRIPTION
**Pull request by kovax**

### CODE REVIEW

File: dsl/src/main/groovy/org/cristalise/dsl/persistency/outcome/OutOfSpecs.groovy [hunk 1]
```diff
@@ -28,5 +28,18 @@ import groovy.transform.CompileStatic;
 @CompileStatic
 class OutOfSpecs {
     String pattern = null
+    String expression = null
     String message = null
+
+    public void setPattern(String p) {
+        if (pattern && expression) throw new InvalidDataException("Do not use pattern and expression at the same time")
```
Comment (taurus227) [inline dsl/src/main/groovy/org/cristalise/dsl/persistency/outcome/OutOfSpecs.groovy]: Why not? Both can raise a warning.

File: dsl/src/main/groovy/org/cristalise/dsl/persistency/outcome/OutOfSpecs.groovy [hunk 2]
```diff
@@ -28,5 +28,18 @@ import groovy.transform.CompileStatic;
...
+
+        pattern = p
+    }
+
+    public void setExpression(String e) {
+        if (pattern && expression) throw new InvalidDataException("Do not use pattern and expression at the same time")
```
Comment (taurus227) [inline dsl/src/main/groovy/org/cristalise/dsl/persistency/outcome/OutOfSpecs.groovy]: Why not? Both can raise a warning.

File: dsl/src/main/groovy/org/cristalise/dsl/persistency/outcome/SchemaDelegate.groovy [hunk 3]
```diff
@@ -178,8 +178,11 @@ class SchemaDelegate {
                     if (f.dynamicForms.updateQuerytRef != null) updateQuerytRef(f.dynamicForms.updateQuerytRef)
                     if (f.dynamicForms.outOfSpecs != null) {
                         outOfSpecs {
-                            if (f.dynamicForms.outOfSpecs.pattern != null) pattern(f.dynamicForms.outOfSpecs.pattern)
-                            if (f.dynamicForms.outOfSpecs.message != null) message(f.dynamicForms.outOfSpecs.message)
+                            if (f.dynamicForms.outOfSpecs.pattern != null)    pattern(f.dynamicForms.outOfSpecs.pattern)
+                            if (f.dynamicForms.outOfSpecs.message != null)    message(f.dynamicForms.outOfSpecs.message)
+                            if (f.dynamicForms.outOfSpecs.expression != null) expression {
+                                mkp.yieldUnescaped("<![CDATA[ ${f.dynamicForms.outOfSpecs.expression}]]>")
```
Comment (taurus227) [inline dsl/src/main/groovy/org/cristalise/dsl/persistency/outcome/SchemaDelegate.groovy]: When is the warning shown? If OutOfSpec.expression evaluates to true, that would tell me that the vaue entered is OutOfSpec, so the warning should be show.

File: dsl/src/main/groovy/org/cristalise/dsl/persistency/outcome/OutOfSpecs.groovy (hunk 1 above)
Comment (kovax) [reply dsl/src/main/groovy/org/cristalise/dsl/persistency/outcome/OutOfSpecs.groovy]: in that case expression shall have a different message.

File: dsl/src/main/groovy/org/cristalise/dsl/persistency/outcome/SchemaDelegate.groovy (hunk 3 above)
Comment (kovax) [reply dsl/src/main/groovy/org/cristalise/dsl/persistency/outcome/SchemaDelegate.groovy]: I do not undrestand what you are saying....

Comment (taurus227) [reply dsl/src/main/groovy/org/cristalise/dsl/persistency/outcome/SchemaDelegate.groovy]: The warning is shown when the expression evaluates to true or when it evaluates to false?

File: dsl/src/main/groovy/org/cristalise/dsl/persistency/outcome/OutOfSpecs.groovy (hunk 1 above)
Comment (kovax) [reply dsl/src/main/groovy/org/cristalise/dsl/persistency/outcome/OutOfSpecs.groovy]: but, I think it is an overkill to specify both OOS pattern and expression for a single field.

File: dsl/src/main/groovy/org/cristalise/dsl/persistency/outcome/SchemaDelegate.groovy (hunk 3 above)
Comment (kovax) [reply dsl/src/main/groovy/org/cristalise/dsl/persistency/outcome/SchemaDelegate.groovy]: I think this requirements is for the webui as it does not effect the functionality of neither  dsl nor outcomebuilder.

Comment (taurus227) [reply dsl/src/main/groovy/org/cristalise/dsl/persistency/outcome/SchemaDelegate.groovy]: OK, we can call it Warning.

**Pull request merged**